src/lib.rs

```diff
@@ -358,20 +388,36 @@ impl CANSocket {
     /// Blocking read a single can frame.
     pub fn read_frame(&self) -> io::Result<CANFrame> {
         let mut frame = CANFrame {
-            _id: 0,
-            _data_len: 0,
-            _pad: 0,
-            _res0: 0,
-            _res1: 0,
-            _data: [0; 8],
+            tag: CANFrameType::Normal,
+
+            /// NOTE: Can't use `std::mem::MaybeUninit::uninit().assume_init()`, which
+            /// would be very appropriate here, because that would result in undefined
+            /// behavior, according to `std::mem::MaybeUninit` docs (1.36.0):
+            ///
+            /// > [...] uninitialized memory is special in that the compiler knows that it
+            /// > does not have a fixed value. This makes it undefined behavior to have
+            /// > uninitialized data in a variable even if that variable has an integer type,
+            /// > which otherwise can hold any fixed bit pattern [...]
+            ///
+            s: unsafe { std::mem::MaybeUninit::zeroed().assume_init() }
         };
 
         let read_rv = unsafe {
-            let frame_ptr = &mut frame as *mut CANFrame;
-            read(self.fd, frame_ptr as *mut c_void, size_of::<CANFrame>())
+            let frame_ptr = &mut frame.s as *mut CANFrameStruct;
+            read(self.fd, frame_ptr as *mut c_void, size_of::<CANFrameStruct>())
         };
 
-        if read_rv as usize != size_of::<CANFrame>() {
+        #[cfg(feature = "can_fd")]
+        if read_rv as usize == CAN_MTU {
```

Comment:
I think this way might be less copying or feature dependent code.

```Rust
frame.tag = match read_rv as usize {
  CAN_MTU => CANFrameType::Normal,
  #[cfg(feature = "can_fd")]
  CANFD_MTU => CANFrameType::Fd,
  _ => return Err(io::Error::last_os_error()),
}
```

Reply:
Sure, this is shorter and much more Rust-idiomatic